src/Castle.Core/DynamicProxy/Contributors/InterfaceMembersWithDefaultImplementationCollector.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright 2004-2022 Castle Project - http://www.castleproject.org/
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy.Contributors
+{
+	using System;
+	using System.Diagnostics;
+	using System.Reflection;
+
+	using Castle.DynamicProxy.Generators;
+
+	internal sealed class InterfaceMembersWithDefaultImplementationCollector : MembersCollector
+	{
+		private readonly InterfaceMapping map;
+
+		public InterfaceMembersWithDefaultImplementationCollector(Type interfaceType, Type classToProxy)
+			: base(interfaceType)
+		{
+			Debug.Assert(interfaceType != null);
+			Debug.Assert(interfaceType.IsInterface);
+
+			Debug.Assert(classToProxy != null);
+			Debug.Assert(classToProxy.IsClass);
+
+			map = classToProxy.GetInterfaceMap(interfaceType);
+		}
+
+		protected override MetaMethod GetMethodToGenerate(MethodInfo method, IProxyGenerationHook hook, bool isStandalone)
+		{
+			var index = Array.IndexOf(map.InterfaceMethods, method);
+			Debug.Assert(index >= 0);
+
+			var methodOnTarget = map.TargetMethods[index];
+			if (methodOnTarget.DeclaringType.IsInterface == false)
+			{
+				return null;
+			}
+
+			var proxyable = AcceptMethod(method, true, hook);
+			if (!proxyable)
+			{
+				return null;
+			}
```

Comment:
It might be possible to optimize this further.

 * The `proxyable = AcceptMethod(...)` check could be run first, which would mean that the `map` lookups do not need to run in the negative case.

 * If `method.IsAbstract == true` (i. e. an interface method _without_ a default implementation), then `methodOnTarget.DeclaringType.IsInterface == false`, because implementing classes are then not allowed to omit declaring that method. So we could check `method.IsAbstract` first and then possibly skip the `map` lookups in the positive case.

Reply:
I've only applied the second optimization; moving up the `AcceptMethod` check didn't seem so crucial anymore afterwards, and all the other collectors usually run that check last, too, so it's nice to stay consistent with them.